ipn/ipnlocal/peerapi.go

```diff
@@ -341,11 +341,71 @@ func (s *peerAPIServer) DeleteFile(baseName string) error {
 // accidentally logging actual filenames anywhere.
 const redacted = "redacted"
 
+type redactedErr struct {
+	msg   string
+	inner error
+}
+
+func (re *redactedErr) Error() string {
+	return re.msg
+}
+
+func (re *redactedErr) Unwrap() error {
+	return re.inner
+}
+
 func redactErr(err error) error {
-	if pe, ok := err.(*os.PathError); ok {
+	var redactStrings []string
+
+	var pe *os.PathError
+	if errors.As(err, &pe) {
+		// If this is the root error, then we can redact it directly.
+		if err == pe {
+			pe.Path = redacted
+			return pe
+		}
+
+		// Otherwise, we have a *PathError somewhere in the error
+		// chain, and we can't redact it because something later in the
+		// chain may have cached the Error() return already (as
+		// fmt.Errorf does).
+		//
+		// Add this path to the set of paths that we will redact, below.
+		redactStrings = append(redactStrings, pe.Path)
```

Comment:
Yuck, gross. It's unfortunate that `fmt.Errorf` caches the error string.

Reply:
Yeah, I was puzzled and then 😵 when I figured out why the tests were failing despite redacting the underlying `PathError` 😢

Reply:
I feel bad since I reviewed that code when it was added to `fmt`. I should have thought about this use-case.

Reply:
All good 😃 This is frankly a pretty niche use-case, TBQH.